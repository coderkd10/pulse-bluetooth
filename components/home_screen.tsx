import { View, Text, Button, FlatList, TouchableOpacity, StyleSheet } from "react-native";
import { BTNotConnectedDevice, BTConnectedDevice, BTReadingDevice } from "@/utils/types";
import { DeviceName } from "./device_name";
import { BTDataReadingModal } from "./bt_data_reading_modal";

type HomeScreenProps = {
  isScanning: boolean,
  handleStartScan: () => void,
  notConnectedDevices: BTNotConnectedDevice[],
  handleConnectDevice: (device: BTNotConnectedDevice) => void,
  connectedDevices: BTConnectedDevice[],
  handleDisconnectDevice: (device: BTConnectedDevice) => void,
  handleReadDevice: (device: BTConnectedDevice) => void,
  currentReadingDevice: BTReadingDevice | null,
  clearCurrentReadingDevice: () => void,
};
export const HomeScreen = (props: HomeScreenProps) => {
  const renderNotConnectedDeviceItem = ({ item: device }: { item: BTNotConnectedDevice }) => {
    const id = device.peripheral.id;
    const name = device.peripheral.name;
    return (<View style={styles.deviceContainer}>
      <DeviceName id={id} name={name} />
      <TouchableOpacity 
        style={[styles.button, device.isConnecting && styles.disabledButton]}
        disabled={device.isConnecting}
        onPress={() => {
          props.handleConnectDevice(device);
        }}
      >
        <Text style={styles.buttonText}>
          {device.isConnecting ? "Connecting" : "Connect"}
        </Text>
      </TouchableOpacity>
    </View>);
  }
  
  const renderConnectedDeviceItem = ({ item: device }: {item: BTConnectedDevice}) => {
    const id = device.peripheral.id;
    const name = device.peripheral.name;
    return (<View style={styles.deviceContainer}>
      <DeviceName id={id} name={name} />
      <View style={styles.connectedDeviceButtonGroup}>
        <TouchableOpacity
          style={[styles.button, styles.redBg]}
          onPress={() => {
            props.handleDisconnectDevice(device);
          }}
        >
          <Text style={styles.buttonText}>Disconnect</Text>
        </TouchableOpacity>
        <TouchableOpacity
          style={[
            styles.button,
            styles.greenBg,
            !device.peripheralInfo && styles.disabledButton
          ]}
          disabled={!device.peripheralInfo}
          onPress={() => {
            props.handleReadDevice(device);
          }}
        >
          <Text style={styles.buttonText}>Read Data</Text>
        </TouchableOpacity>
      </View>
    </View>);
  }

  return (
    <View style={styles.container}>
      <Button 
        title={props.isScanning ? "Scanning ...": "Scan Bluetooth Devices"} 
        onPress={() => props.handleStartScan()}
        disabled={props.isScanning}
      />
      <Text style={styles.subtitle}>Discovered Devices ({props.notConnectedDevices.length}):</Text>
      <FlatList
        data={props.notConnectedDevices}
        renderItem={renderNotConnectedDeviceItem}
        keyExtractor={device => device.peripheral.id}
        style={styles.list}
      />
      <Text style={styles.subtitle}>Connected Devices ({props.connectedDevices.length}):</Text>
      <FlatList
        data={props.connectedDevices}
        renderItem={renderConnectedDeviceItem}
        keyExtractor={device => device.peripheral.id}
        style={styles.list}
      />
      <BTDataReadingModal 
        currentReadingDevice={props.currentReadingDevice}
        handleClose={() => {
            props.clearCurrentReadingDevice();
        }}
      />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 20,
    backgroundColor: '#fff',
  },
  subtitle: {
    fontSize: 18,
    fontWeight: 'bold',
    marginVertical: 10,
  },
  list: {
    padding: 5,
    paddingBottom: 100,
    marginBottom: 20,
    borderWidth: 1,
    borderRadius: 5,
    minHeight: '10%'
  },
  deviceContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    padding: 10,
    marginBottom: 10,
    borderWidth: 1,
    borderColor: '#ccc',
    borderRadius: 5,
  },
  button: {
    backgroundColor: '#007BFF',
    padding: 10,
    borderRadius: 5,
    width: 100,
    alignItems: 'center'
  },
  buttonText: {
    color: '#fff',
    fontWeight: 'bold',
  },
  disabledButton: {
    backgroundColor: '#ccc',
  },
  greenBg: {
    backgroundColor: 'green',
  },
  redBg: {
    backgroundColor: 'red',
  },
  connectedDeviceButtonGroup: {
    flexDirection: 'row',
    gap: 10,
  },
});
