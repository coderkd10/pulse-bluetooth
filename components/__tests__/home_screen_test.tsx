import * as React from 'react';
import { render } from '@testing-library/react-native';

import { HomeScreen } from '../home_screen';

const props = {
    isScanning: false,
    handleStartScan: jest.fn(),
    notConnectedDevices: [
        {
            peripheral: {
                id: "62:64:0D:5F:87:9C",
                name: "JBL Flip",
                rssi: 0,
                advertising: {},
            },
            isConnecting: false,
        },
        {
            peripheral: {
                id: "34:4F:B1:C5:3E:D9",
                name: "My BT Speaker",
                rssi: 0,
                advertising: {},
            },
            isConnecting: true,
        },
    ],
    handleConnectDevice: jest.fn(),
    connectedDevices: [
        {
            peripheral: {
                id: "75:8C:9C:43:EE:4C",
                name: "Pulse BT",
                rssi: 0,
                advertising: {},
            },
        }
    ],
    handleDisconnectDevice: jest.fn(),
    handleReadDevice: jest.fn(),
    currentReadingDevice: null,
    clearCurrentReadingDevice: jest.fn(),
};

it(`renders correctly`, () => {
    const tree = render(
        <HomeScreen {...props} />
    ).toJSON();
  
    expect(tree).toMatchSnapshot();
});
  