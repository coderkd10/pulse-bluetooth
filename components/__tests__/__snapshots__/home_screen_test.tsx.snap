// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "#fff",
      "flex": 1,
      "padding": 20,
    }
  }
>
  <View
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": false,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        [
          {},
        ]
      }
    >
      <Text
        disabled={false}
        style={
          [
            {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Scan Bluetooth Devices
      </Text>
    </View>
  </View>
  <Text
    style={
      {
        "fontSize": 18,
        "fontWeight": "bold",
        "marginVertical": 10,
      }
    }
  >
    Discovered Devices (
    2
    ):
  </Text>
  <RCTScrollView
    data={
      [
        {
          "isConnecting": false,
          "peripheral": {
            "advertising": {},
            "id": "62:64:0D:5F:87:9C",
            "name": "JBL Flip",
            "rssi": 0,
          },
        },
        {
          "isConnecting": true,
          "peripheral": {
            "advertising": {},
            "id": "34:4F:B1:C5:3E:D9",
            "name": "My BT Speaker",
            "rssi": 0,
          },
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    style={
      {
        "borderRadius": 5,
        "borderWidth": 1,
        "marginBottom": 20,
        "minHeight": "10%",
        "padding": 5,
        "paddingBottom": 100,
      }
    }
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "alignItems": "center",
              "borderColor": "#ccc",
              "borderRadius": 5,
              "borderWidth": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginBottom": 10,
              "padding": 10,
            }
          }
        >
          <View>
            <Text
              style={
                {
                  "fontSize": 16,
                }
              }
            >
              JBL Flip
            </Text>
            <Text
              style={
                {
                  "color": "#666",
                  "fontWeight": "bold",
                }
              }
            >
              62:64:0D:5F:87:9C
            </Text>
          </View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": false,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#007BFF",
                "borderRadius": 5,
                "opacity": 1,
                "padding": 10,
                "width": 100,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#fff",
                  "fontWeight": "bold",
                }
              }
            >
              Connect
            </Text>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "alignItems": "center",
              "borderColor": "#ccc",
              "borderRadius": 5,
              "borderWidth": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginBottom": 10,
              "padding": 10,
            }
          }
        >
          <View>
            <Text
              style={
                {
                  "fontSize": 16,
                }
              }
            >
              My BT Speaker
            </Text>
            <Text
              style={
                {
                  "color": "#666",
                  "fontWeight": "bold",
                }
              }
            >
              34:4F:B1:C5:3E:D9
            </Text>
          </View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": true,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#ccc",
                "borderRadius": 5,
                "opacity": 1,
                "padding": 10,
                "width": 100,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#fff",
                  "fontWeight": "bold",
                }
              }
            >
              Connecting
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <Text
    style={
      {
        "fontSize": 18,
        "fontWeight": "bold",
        "marginVertical": 10,
      }
    }
  >
    Connected Devices (
    1
    ):
  </Text>
  <RCTScrollView
    data={
      [
        {
          "peripheral": {
            "advertising": {},
            "id": "75:8C:9C:43:EE:4C",
            "name": "Pulse BT",
            "rssi": 0,
          },
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    style={
      {
        "borderRadius": 5,
        "borderWidth": 1,
        "marginBottom": 20,
        "minHeight": "10%",
        "padding": 5,
        "paddingBottom": 100,
      }
    }
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "alignItems": "center",
              "borderColor": "#ccc",
              "borderRadius": 5,
              "borderWidth": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginBottom": 10,
              "padding": 10,
            }
          }
        >
          <View>
            <Text
              style={
                {
                  "fontSize": 16,
                }
              }
            >
              Pulse BT
            </Text>
            <Text
              style={
                {
                  "color": "#666",
                  "fontWeight": "bold",
                }
              }
            >
              75:8C:9C:43:EE:4C
            </Text>
          </View>
          <View
            style={
              {
                "flexDirection": "row",
                "gap": 10,
              }
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "red",
                  "borderRadius": 5,
                  "opacity": 1,
                  "padding": 10,
                  "width": 100,
                }
              }
            >
              <Text
                style={
                  {
                    "color": "#fff",
                    "fontWeight": "bold",
                  }
                }
              >
                Disconnect
              </Text>
            </View>
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": true,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "alignItems": "center",
                  "backgroundColor": "#ccc",
                  "borderRadius": 5,
                  "opacity": 1,
                  "padding": 10,
                  "width": 100,
                }
              }
            >
              <Text
                style={
                  {
                    "color": "#fff",
                    "fontWeight": "bold",
                  }
                }
              >
                Read Data
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
